build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/TestingConventionsCheckTask.java

```diff
@@ -0,0 +1,284 @@
+            return testClassesCandidate.stream()
+                .filter(testClassDefaultPredicate)
+                .filter(extendsBaseClass)
+                .filter(TestingConventionsCheckWorkAction::seemsLikeATest)
+                .collect(Collectors.toList());
+        }
+
+        private static boolean seemsLikeATest(Class<?> clazz) {
+            try {
+                Class<?> junitTest = loadClassWithoutInitializing("org.junit.Assert", clazz.getClassLoader());
+                if (junitTest.isAssignableFrom(clazz)) {
+                    Logging.getLogger(TestingConventionsCheckWorkAction.class)
+                        .debug("{} is a test because it extends {}", clazz.getName(), junitTest.getName());
+                    return true;
+                }
+
+                Class<?> junitAnnotation = loadClassWithoutInitializing("org.junit.Test", clazz.getClassLoader());
+                for (Method method : clazz.getMethods()) {
+                    if (matchesTestMethodNamingConvention(method)) {
+                        Logging.getLogger(TestingConventionsCheckWorkAction.class)
+                            .debug("{} is a test because it has method named '{}'", clazz.getName(), method.getName());
+                        return true;
+                    }
+                    if (isAnnotated(method, junitAnnotation)) {
+                        Logging.getLogger(TestingConventionsCheckWorkAction.class)
+                            .debug(
+                                "{} is a test because it has method '{}' annotated with '{}'",
+                                clazz.getName(),
+                                method.getName(),
+                                junitAnnotation.getName()
+                            );
+                        return true;
+                    }
+                }
+
+                return false;
+            } catch (NoClassDefFoundError e) {
+                // Include the message to get more info to get more a more useful message when running Gradle without -s
+                throw new IllegalStateException("Failed to inspect class " + clazz.getName() + ". Missing class? " + e.getMessage(), e);
+            }
+        }
+
+        private static boolean matchesTestMethodNamingConvention(Method method) {
+            return method.getName().startsWith(JUNIT3_TEST_METHOD_PREFIX)
+                && Modifier.isStatic(method.getModifiers()) == false
+                && method.getReturnType().equals(Void.TYPE);
+        }
+
+        private static boolean isAnnotated(Method method, Class<?> annotation) {
+            return List.of(method.getAnnotations())
+                .stream()
+                .anyMatch(presentAnnotation -> annotation.isAssignableFrom(presentAnnotation.getClass()));
+        }
+
+        private List<String> walkPathAndLoadClasses(File testRoot) {
```

Comment:
Couldn't we simplify this by using `FileTree.visit()` and leveraging `RelativePath`?

Reply:
just copied this over from the original class. yeah I simplified this now